src/ring_theory/ideal/norm.lean

```diff
@@ -59,3 +67,187 @@ add_subgroup.index_eq_one.trans (by simp [set_like.ext_iff])
+This is essentially just a repackaging of the Chinese Remainder Theorem.
+-/
+lemma card_quot_mul_of_coprime [is_dedekind_domain S] [module.free ℤ S] [module.finite ℤ S]
+  (I J : ideal S) (coprime : I ⊔ J = ⊤) : card_quot (I * J) = card_quot I * card_quot J :=
+begin
+  let b := module.free.choose_basis ℤ S,
+  casesI is_empty_or_nonempty (module.free.choose_basis_index ℤ S),
+  { haveI : subsingleton S := function.surjective.subsingleton b.repr.to_equiv.symm.surjective,
+    nontriviality S,
+    exfalso,
+    exact not_nontrivial_iff_subsingleton.mpr ‹subsingleton S› ‹nontrivial S› },
+  haveI : infinite S := infinite.of_surjective _ b.repr.to_equiv.surjective,
+  by_cases hI : I = ⊥,
+  { rw [hI, submodule.bot_mul, card_quot_bot, zero_mul] },
+  by_cases hJ : J = ⊥,
+  { rw [hJ, submodule.mul_bot, card_quot_bot, mul_zero] },
+  have hIJ : I * J ≠ ⊥ := mt ideal.mul_eq_bot.mp (not_or hI hJ),
+
+  letI := classical.dec_eq (module.free.choose_basis_index ℤ S),
+  letI := I.fintype_quotient_of_free_of_ne_bot hI,
+  letI := J.fintype_quotient_of_free_of_ne_bot hJ,
+  letI := (I * J).fintype_quotient_of_free_of_ne_bot hIJ,
+
+  rw [card_quot_apply, card_quot_apply, card_quot_apply,
+      fintype.card_eq.mpr ⟨(ideal.quotient_mul_equiv_quotient_prod I J coprime).to_equiv⟩,
+      fintype.card_prod]
+end
+
+/-- If `a ∈ P^i \ P^(i+1) c ∈ P^i`, then `a * d + e = c` for `e ∈ P^(i+1)`.
+`ideal.mul_add_mem_pow_succ_unique` shows the choice of `d` is unique, up to `P`.
+Inspired by [Neukirch], proposition 6.1 -/
+lemma ideal.exists_mul_add_mem_pow_succ [is_dedekind_domain S]
+  (P : ideal S) [P_prime : P.is_prime] (hP : P ≠ ⊥) {i : ℕ}
+  (a c : S) (a_mem : a ∈ P ^ i) (a_not_mem : a ∉ P ^ (i + 1)) (c_mem : c ∈ P ^ i) :
+  ∃ (d : S) (e ∈ P ^ (i + 1)), a * d + e = c :=
+begin
+  suffices eq_b : P ^ i = ideal.span {a} ⊔ P ^ (i + 1),
+  { rw eq_b at c_mem,
+    simp only [mul_comm a],
+    exact ideal.mem_span_singleton_sup.mp c_mem },
+  refine (ideal.eq_prime_pow_of_succ_lt_of_le hP
+    (lt_of_le_of_ne le_sup_right _)
+    (sup_le (ideal.span_le.mpr (set.singleton_subset_iff.mpr a_mem))
+      (ideal.pow_succ_lt_pow hP i).le)).symm,
+  contrapose! a_not_mem with this,
+  rw this,
+  exact mem_sup.mpr ⟨a, mem_span_singleton_self a, 0, by simp, by simp⟩
+end
+
+lemma prime_pow_succ_dvd_mul {α : Type*} [cancel_comm_monoid_with_zero α]
+  {p x y : α} (h : prime p) {i : ℕ} (hxy : p ^ (i + 1) ∣ x * y) :
+  p ^ (i + 1) ∣ x ∨ p ∣ y :=
+begin
+  rw or_iff_not_imp_right,
+  intro hy,
+  induction i with i ih generalizing x,
+  { simp only [zero_add, pow_one] at *,
+    exact (h.dvd_or_dvd hxy).resolve_right hy },
+  rw pow_succ at hxy ⊢,
+  obtain ⟨x', rfl⟩ := (h.dvd_or_dvd (dvd_of_mul_right_dvd hxy)).resolve_right hy,
+  rw mul_assoc at hxy,
+  exact mul_dvd_mul_left p (ih ((mul_dvd_mul_iff_left h.ne_zero).mp hxy)),
+end
+
+lemma ideal.mem_prime_of_mul_mem_pow [is_dedekind_domain S]
+  {P : ideal S} [P_prime : P.is_prime] (hP : P ≠ ⊥) {i : ℕ}
```

Comment:
I think you can declare `{P : ideal S}` as a variable, since it is used in several declarations (maybe even with some other assumptions on it).